src/Data/AudienceTree.elm

```diff
@@ -0,0 +1,361 @@
+module Data.AudienceTree exposing
+    ( AudienceTree
+    , DuplicatedIdFolder
+    , Errors
+    , audiences
+    , children
+    , fromAudiencesAndFolders
+    , name
+    )
+
+import Data.Audience exposing (Audience)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Dict exposing (Dict)
+
+
+type AudienceTree
+    = AudienceTree AudienceTreeInternal
+
+
+type alias AudienceTreeInternal =
+    { name : String
+    , children : List AudienceTree
+    , audiences : List String
+    }
+
+
+children : AudienceTree -> List AudienceTree
+children (AudienceTree a) =
+    a.children
+
+
+name : AudienceTree -> String
+name (AudienceTree a) =
+    a.name
+
+
+audiences : AudienceTree -> List String
+audiences (AudienceTree a) =
+    a.audiences
+
+
+type alias Errors =
+    { duplicatedFolderIds : List DuplicatedIdFolder
+    , missingFolderId : List MissingParentAudience
+    , missingFolderParentId : List MissingParentFolder
+    }
+
+
+{-| Build the audience tree from the lists of audiences and audience folders.
+It will always succeed to build some Tree, even if some ids are inconsistent.
+E.g. if an audience folder references a non-existing folder, this audience is directly
+added to the root, and this audience is added to the `missingFolderId` field.
+-}
+fromAudiencesAndFolders : List Audience -> List AudienceFolder -> ( AudienceTree, Errors )
+fromAudiencesAndFolders audiences_ folders =
+    let
+        step3result =
+            step1 folders
+                |> step2 audiences_
+                |> step3
```

Comment:
I don't like those generic names `stepN` and even fact, that those function are defined below their usage, but I don't have better naming for you know and "function definitions below it's first usage" is something I fighting with even in our production code :D

Reply:
Having the function defined before it's usage is something I would like to see more in our codebase as well, it is easily searchable this way in Vim for me :)

Reply:
Not this flamewar again :D

Reply:
I don't like those names neither... I searched for better names but didn't succeed and finally decided it was private things and it wasn't a big deal (yeah I'm lazy sometime...).

Reply:
And about "function definitions above it's first usage", I'd say it is an historical technical constraint. My thinking about this is "first describe the problem at high level, then give implementation details". 

That being said, with the "Go To Definition" feature accessible with one key press, I think declaration order doesn't really matter since with large files, we rarely read them linearly.

Reply:
It's not entirely true and this approach is forcing other to use advanced IDE tools as you are using. For you order doesn't matter so is that big issue follow rule "declaration first" which could help others? And also think about code review in GH interface where is usual to follow code from top to bottom by just scrolling and jumping from top to bottom and back during code review is something annoying, trust me :) 
But yeah, stop this flame, this is not the place for it, peace 🦄 and 🌈. Also 🥦 for all.